### PR TITLE
Perceiver interpolate position embedding

### PR DESCRIPTION
**Pull request by justyx404**
# What does this PR do?

Add interpolate position embedding to Perceiver #30579. Not really sure how to test the correctness of output value since it does change the entire forward pass. However, correctness of output shape is pretty straight forward.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?
Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@amyeroberts

### CODE REVIEW

File: src/transformers/models/perceiver/modeling_perceiver.py [hunk 1]
```diff
@@ -2749,9 +2765,30 @@ def num_dimensions(self) -> int:
     def output_size(self, *args, **kwargs) -> int:
         return self._num_channels
 
-    def forward(self, batch_size: int) -> torch.Tensor:
+    def interpolate_pos_encoding(self, position_embeddings: torch.Tensor, height: int, width: int) -> torch.Tensor:
+        num_positions, dim = position_embeddings.shape[0], position_embeddings.shape[1]
```
Comment (amyeroberts) [inline src/transformers/models/perceiver/modeling_perceiver.py:2769]: Could we use a more descriptive name than `dim` here? e.g. hidden_dim or whatever it's supposed to represent. All dims are dims :)

File: src/transformers/models/perceiver/modeling_perceiver.py [hunk 2]
```diff
@@ -2749,9 +2765,30 @@ def num_dimensions(self) -> int:
...
+        position_embeddings = position_embeddings.reshape(
+            1, int(math.sqrt(num_positions)), int(math.sqrt(num_positions)), dim
+        ).permute(0, 3, 1, 2)
+        position_embeddings = nn.functional.interpolate(
+            position_embeddings,
+            scale_factor=(height / math.sqrt(num_positions), width / math.sqrt(num_positions)),
+            mode="bicubic",
+            align_corners=False,
+        )
+        position_embeddings = position_embeddings.reshape(1, dim, -1).permute(0, 2, 1).squeeze(0)
```
Comment (amyeroberts) [inline src/transformers/models/perceiver/modeling_perceiver.py:2779]: Let's just calculate the sqrt once 

```suggestion
        new_height = new_width = math.sqrt(num_positions)
        position_embeddings = position_embeddings.reshape(
            1, int(new_height), int(new_width), dim
        ).permute(0, 3, 1, 2)
        position_embeddings = nn.functional.interpolate(
            position_embeddings,
            scale_factor=(height / new_height, width / new_width,
            mode="bicubic",
            align_corners=False,
        )
        position_embeddings = position_embeddings.reshape(1, dim, -1).permute(0, 2, 1).squeeze(0)
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this! 

Just a few small comments

File: src/transformers/models/perceiver/modeling_perceiver.py (hunk 1 above)
Comment (g1y5x3) [reply src/transformers/models/perceiver/modeling_perceiver.py:2769]: How about `self._num_channels`? after all they are the same thing.

File: src/transformers/models/perceiver/modeling_perceiver.py [hunk 3]
```diff
@@ -2859,7 +2896,13 @@ def __init__(self, config: PerceiverConfig) -> None:
     def num_channels(self) -> int:
         return self.config.d_model
 
-    def forward(self, inputs: torch.LongTensor, pos: Optional[torch.Tensor] = None, network_input_is_1d: bool = True):
+    def forward(
+        self,
+        inputs: torch.LongTensor,
+        pos: Optional[torch.Tensor] = None,
+        network_input_is_1d: bool = True,
+        interpolate_pos_encoding: bool = None,
```
[the remark this answers is not in the document]
Comment (g1y5x3) [reply src/transformers/models/perceiver/modeling_perceiver.py]: When I ran the test, `TextPreprocesor` and `AudioPreprocessor` didn't fail but only `MutimodalPreprocessor` did so I missed those. Just added.

**Comment by g1y5x3** [comment]
Thanks for the review, I have addressed the comments. Turns out after working on the no longer needed `DETR`, it makes sorting through the code path of a different model much faster.

File: src/transformers/models/perceiver/modeling_perceiver.py [hunk 4]
```diff
@@ -2859,7 +2897,13 @@ def __init__(self, config: PerceiverConfig) -> None:
     def num_channels(self) -> int:
         return self.config.d_model
 
-    def forward(self, inputs: torch.LongTensor, pos: Optional[torch.Tensor] = None, network_input_is_1d: bool = True):
+    def forward(
+        self,
+        inputs: torch.LongTensor,
+        pos: Optional[torch.Tensor] = None,
+        network_input_is_1d: bool = True,
+        interpolate_pos_encoding: bool = False,
```
Comment (amyeroberts) [inline src/transformers/models/perceiver/modeling_perceiver.py:2905]: `interpolate_pos_encoding` still doesn't appear to be used in this method

File: src/transformers/models/perceiver/modeling_perceiver.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/models/perceiver/modeling_perceiver.py]: `interpolate_pos_encoding` still isn't used in this method., or the one for `PerceiverAudioPreprocessor`. Is this for providing a consistent signature?

File: src/transformers/models/perceiver/modeling_perceiver.py [hunk 5]
```diff
@@ -857,7 +869,9 @@ def forward(
         return_dict = return_dict if return_dict is not None else self.config.use_return_dict
 
         if self.input_preprocessor is not None:
-            inputs, modality_sizes, inputs_without_pos = self.input_preprocessor(inputs)
+            inputs, modality_sizes, inputs_without_pos = self.input_preprocessor(
+                inputs, interpolate_pos_encoding=interpolate_pos_encoding
+            )
```
Comment (g1y5x3) [inline src/transformers/models/perceiver/modeling_perceiver.py:874]: @amyeroberts since `self.input_processor` here could also be either `TextPreprocessor` or `MultimodalPreprosessor`, it would complain `got unexpected keyword argument` if it was not added to the `forward()` of those two corresponding methods. Alternatively, we could add another condition here to call `self.input_processor` with two arguments unless it is `ImagePreprocessor`

Comment (amyeroberts) [reply src/transformers/models/perceiver/modeling_perceiver.py:874]: No, that's OK, I just wanted to make sure that this was deliberate

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this!
